RobustTrees/RobustTreeUtils.py

```diff
@@ -0,0 +1,344 @@
+        curr_path = terminal_path_dict[j]
+        curr_cost = master.eta * int(curr_v) # Start with cost if correctly classify point
+        best_so_far = True
+        for x in range(len(curr_path) - 1):
+            n = curr_path[x] # Current node
+            f = curr_features[x]
+            theta = curr_cutoffs[x]
+            min_f = curr_mins[f]
+            max_f = curr_maxes[f]
+
+            curr_value = master.X.at[i, f] + curr_xi[f]
+            # Went right
+            if (2 * n) + 1 == curr_path[x + 1]: # Path goes right
+                if curr_value <= theta:
+                    # See if can switch to go right by increasing x to theta+1
+                    if max_f < theta+1:
+                        # Impossible path
+                        best_so_far = False
+                        break
+                    # x + delta_x = theta + 1
+                    delta_x = theta - master.X.at[i, f] + 1 # positive value
+
+                    # cost increases by gamma per unit increase of xi
+                    curr_cost += master.gammas.loc[i][f] * (delta_x - curr_xi[f])
+                    curr_xi[f] = delta_x
+
+                # Update bounds
+                curr_mins[f] = max(curr_mins[f], theta + 1)
+
+            else: # Went left
+                if curr_value >= theta+1:
+                    # See if can switch to go left by decreasing x to theta
+                    if min_f > theta:
+                        # Impossible path
+                        best_so_far = False
+                        break
+                    # x + delta_x = theta
+                    delta_x = theta - master.X.at[i, f] # negative value
+
+                    # cost increases by gamma per unit decrease of xi
+                    curr_cost += master.gammas.loc[i][f] * (curr_xi[f] - delta_x)
+                    curr_xi[f] = delta_x
+                    
+                # Update bounds
+                curr_maxes[f] = min(curr_maxes[f], theta)
+
+            if curr_cost > best_cost:
+                # No need to go further
+                best_so_far = False
+                break
+        if best_so_far:
+            best_cost = curr_cost
+            best_path = curr_path
+            xi = curr_xi
+            v = curr_v
+
+    return best_path, best_cost, xi, v
+
+def subproblem(master, i, terminal_nodes, terminal_path_dict, terminal_features_dict, 
+    terminal_assignments_dict, terminal_cutoffs_dict, initial_xi={}, initial_mins={}, initial_maxes={}):
+    label_i = master.y[i]
+    target = [] # list of nodes whose edge to the sink is part of the min cut
+
+    # Solve shortest path problem via DFS w/ pruning
+    target, cost, xi, v = shortest_path_solver(master, i, label_i, terminal_nodes, terminal_path_dict, terminal_features_dict, 
+        terminal_assignments_dict, terminal_cutoffs_dict, initial_xi, initial_mins, initial_maxes)
+    
+    return target, xi, cost, v
+
+def mycallback(model, where):
```

Comment:
Do we need this callback? My understanding is that this was mostly used for the paper replication code

Reply:
Yes, so we do need the callback to feed to the gurobi solver in the fitting step since Robust Trees can only be solved using a Benders approach with callbacks